File: server/swift/Sources/App/Domain/ValueObjects/ItemImage.swift
import Foundation

/// Item image value object
struct ItemImage: Codable, Equatable, Sendable {
    let url: String
    let alt: String?
    let sortOrder: Int?

    enum CodingKeys: String, CodingKey {
        case url
        case alt
        case sortOrder = "sort_order"
    }

    init(url: String, alt: String? = nil, sortOrder: Int? = nil) {
        self.url = url
        self.alt = alt
        self.sortOrder = sortOrder
    }
}

File: server/go/internal/domain/valueobject/item_image.go
package valueobject

import (
	"encoding/json"
)

// ItemImage represents an image associated with an item
type ItemImage struct {
	URL       string `json:"url" binding:"required,url"`
	Alt       string `json:"alt,omitempty"`
	SortOrder int    `json:"sortOrder,omitempty"`
}

// MarshalJSON implements json.Marshaler
func (ii ItemImage) MarshalJSON() ([]byte, error) {
	type alias ItemImage
	return json.Marshal(struct {
		URL       string `json:"url"`
		Alt       string `json:"alt,omitempty"`
		SortOrder int    `json:"sortOrder,omitempty"`
	}{
		URL:       ii.URL,
		Alt:       ii.Alt,
		SortOrder: ii.SortOrder,
	})
}

// UnmarshalJSON implements json.Unmarshaler
func (ii *ItemImage) UnmarshalJSON(data []byte) error {
	type alias ItemImage
	aux := &struct {
		URL       string `json:"url" binding:"required,url"`
		Alt       string `json:"alt,omitempty"`
		SortOrder *int   `json:"sortOrder,omitempty"`
	}{
		URL: ii.URL,
		Alt: ii.Alt,
	}
	if err := json.Unmarshal(data, &aux); err != nil {
		return err
	}
	ii.URL = aux.URL
	ii.Alt = aux.Alt
	if aux.SortOrder != nil {
		ii.SortOrder = *aux.SortOrder
	}
	return nil
}
